src/dodal/common/beamlines/instantiation_behaviour.py

```diff
@@ -0,0 +1,131 @@
+from collections.abc import Callable
+from dataclasses import dataclass
+
+from bluesky.run_engine import call_in_bluesky_event_loop
+from ophyd_async.core import DEFAULT_TIMEOUT
+from ophyd_async.core import Device as OphydV2Device
+
+from dodal.common.beamlines.beamline_utils import (
+    ACTIVE_DEVICES,
+)
+from dodal.utils import AnyDevice
+
+
+class XYZDetector(OphydV2Device):
+    def __init__(self, prefix: str, *args, **kwargs):
+        self.prefix = prefix
+        super().__init__(*args, **kwargs)
+
+    @property
+    def hints(self):
+        raise NotImplementedError
+
+
+@dataclass
+class DeviceInitializationConfig:
+    """
+    eager ones all connect at the beginning in blueapi startup. if they fail to connect we get an error
+    endpoint for what devices are failing
+    only matters in connect failure reporting behavior
+    lazy and eager are not about when connect but whether should be connected
+    """
+
+    eager: bool = True
+    set_name: bool = True
+    default_timeout_for_connect: float = DEFAULT_TIMEOUT
+    default_use_mock_at_connection: bool = False
+
+
+class DeviceInitializationController:
+    """
+    This class is responsible for the instantiation of a device.
+    The device is then cached here locally, and also in the global variable ACTIVE_DEVICES.
+    """
+
+    device: AnyDevice | None = None
+    factory: Callable[[], AnyDevice]
+
+    def __init__(
+        self, config: DeviceInitializationConfig, factory: Callable[[], AnyDevice]
+    ) -> None:
+        self.factory = factory
+        self.config = config
+        CONTROLLERS[self.factory.__name__] = self
+        self.device = None
+
+    def see_if_device_is_in_cache(self) -> AnyDevice | None:
+        return self.device
+
+    def __call__(
+        self,
+        connect=False,
+        mock: bool | None = None,
+        timeout: float | None = None,
+        name: str | None = None,
+    ) -> AnyDevice:
+        device = self.see_if_device_is_in_cache() or self.factory()
+        if self.config.set_name:
+            device.set_name(self.factory.__name__)
+        self.add_device_to_cache(device)
+
+        if connect:
+            call_in_bluesky_event_loop(
+                device.connect(
+                    timeout=self.config.default_timeout_for_connect
+                    if timeout is None
+                    else timeout,
+                    mock=self.config.default_use_mock_at_connection
+                    if mock is None
+                    else mock,
+                )
+            )
+            assert self.device is not None
+            CONTROLLERS[self.device.name] = self
+
+        if self.device is not None:
+            CONTROLLERS[self.device.name] = self
+        return device
+
+    def add_device_to_cache(self, device: AnyDevice) -> None:
+        self.device = device
+        ACTIVE_DEVICES[device.name] = device
+
+
+CONTROLLERS: dict[str, DeviceInitializationController] = {}
+
+
+def instantiation_behaviour(
```

Comment:
Should: Apologies if there's already been a discussion around it, but can we change this name? I'm concerned because `_behaviour` is a redundant suffix (see point 8 [here](https://medium.com/krinati-engineering/clean-code-meaningful-naming-conventions-e3a161d90745)) and because it didn't make sense to me until I read a function it was decorating. Some alternatives:

1. `device_instantiation`
2. `device_factory`
3. `device_configuration`
4. `device_singleton`

Reply:
`device_instantiation` is bad for a function name here, as `device_instantiation` is an _event_ that happens whenever we call the `waxs()` (where waxs = some_device), so a function with that name is confusing
`factory` is technically true but it's an oop statement fitting for a class, but here it is a decorator
`device_configuration` is underspecified - this is specifically `device_iniitalization_configuration`
`singleton` is inaccurate as it does not prevent by itself more than 1 instance afaik

`add_instantiation_behavior_to_device` would be my suggestion.
Zen of Python says:
- Explicit is better than implicit.
- [Sparse](https://en.wikipedia.org/wiki/Minimalism#Design,_architecture,_and_spaces) is better than dense.
- [Readability](https://en.wikipedia.org/wiki/Computer_programming#Readability_of_source_code) counts.

Reply:
@DominicOram what is your take on this?

Reply:
My preference would be `device_factory`. I like that it's also effectively helping to label the function as a factory too.

Reply:
It also matches the name in the issue and the PR

Reply:
done